core/modules/qdisp/LargeResultMgr.cc

```diff
@@ -0,0 +1,142 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+// Class header
+#include "qdisp/LargeResultMgr.h"
+
+// System headers
+
+// Third-party headers
+#include "XrdSsi/XrdSsiRequest.hh"
+
+// LSST headers
+#include "lsst/log/Log.h"
+
+// Qserv headers
+
+namespace {
+LOG_LOGGER _log = LOG_GET("lsst.qserv.qdisp.LargeResultMgr");
+
+}
+
+namespace lsst {
+namespace qserv {
+namespace qdisp {
+
+
+/// Set the value of the xrootd semaphore for tracking large response blocks to start.
+void LargeResultMgr::_setup() {
+    std::lock_guard<std::mutex> lck(_mtx);
+    for (int j = 0; j < _runningCountMax; ++j) {
+        _post("initialization");
+    }
+}
+
+
+/// Check if any large result blocks can be run.
+void LargeResultMgr::startBlock(std::string const& jobId) {
+    std::lock_guard<std::mutex> lck(_mtx);
+    ++_blockCount;
+    LOGS(_log, LOG_LVL_DEBUG, jobId << " LargeResultMgr::start blockCount=" << _blockCount);
+}
+
+
+/// Decrement the number of running blocks and see if any can be started
+void LargeResultMgr::finishBlock(std::string const& jobId) {
+    std::lock_guard<std::mutex> lck(_mtx);
+    --_blockCount;
+    if (_delayAllPosts) {
+        _delayPost(jobId);
+    } else {
+        _post(jobId);
+        _freeDelayedPosts();
+    }
+}
+
+
+/// Increase the count of user queries going out and possible delay large results.
+void LargeResultMgr::incrOutGoingQueries() {
+    std::lock_guard<std::mutex> lck(_mtx);
+    ++_outGoingQueries;
+    LOGS(_log, LOG_LVL_DEBUG, "incrOutGoingQueries outGoingQueries=" << _outGoingQueries);
+    _setDelayAllPosts(true);
+}
+
+
+/// Decrease the count of outgoing queries and possibly allow incoming large results to continue.
+void LargeResultMgr::decrOutGoingQueries() {
+    std::lock_guard<std::mutex> lck(_mtx);
+    --_outGoingQueries;
+    LOGS(_log, LOG_LVL_DEBUG, "decrOutGoingQueries outGoingQueries=" << _outGoingQueries);
+    // Allow some large result processing whenever a query finishes being sent out.
+    _freeDelayedPosts();
+    _setDelayAllPosts(_outGoingQueries > 0);
+}
+
+
+/// If delayAll is true, all incoming large results will be delayed
+/// Must hold _mtx when calling this function.
+void LargeResultMgr::_setDelayAllPosts(bool delayAll) {
+    if (delayAll == _delayAllPosts) return;
+    _delayAllPosts = delayAll;
+    LOGS(_log, LOG_LVL_INFO, " change delayAll=" << _delayAllPosts);
+    if (!_delayAllPosts) {
+        _freeDelayedPosts();
+    }
+}
+
+
+/// Increment xrootd's large result semaphore to allow another large result to run.
```

Comment:
Just a nit pick here. mentions of xrootd should be replaced by SSI. Note that the publicframework does not have a stick of xrootd code in it. Indeed, the semaphore is specific to SSI and xrootd does not get involved whatsoever here.